app/waf/src/controllers/declaration.controller.ts

```diff
@@ -423,4 +432,85 @@ export class DeclarationController extends BaseController {
       },
     );
   }
+
+  async assertVirtualAddressAvailability(
+    adc: Adc,
+    declaration: AS3Declaration,
+  ): Promise<void> {
+    try {
+      let netId = (() => {
+        for (let netName of Object.keys(adc.networks)) {
+          let net = adc.networks[netName];
+          if (net.type === 'ext') return net.networkId;
+        }
+        throw new Error(
+          `not found the external network from adc: ${adc.networks}`,
+        );
+      })();
+
+      let netHelper = await this.wafapp.get(WafBindingKeys.KeyNetworkDriver);
+
+      let userToken = await this.reqCxt.get(
+        WafBindingKeys.Request.KeyUserToken,
+      );
+      let ports = await netHelper.getPortsInfo(userToken, netId);
```

Comment:
getting all ports of this network then checking each one is too heavy. we need to find a better way.

Reply:
Or do you know how to query a port with query parameter of fixed_ips array?

Reply:
if my proposal is to be adopted, we don't need to "query a port with parameter of fixed_ips array" anymore. Also, we might be able to delete many codes, e.g. getPorts info etc.